kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/utils/MessageIdUtils.java

```diff
@@ -91,4 +106,66 @@ public static final long offsetAfterBatchIndex(long offset) {
         }
         return offset;
     }
+
+    public static long getCurrentOffset(ManagedLedger managedLedger) {
+        return ((ManagedLedgerInterceptorImpl) managedLedger.getManagedLedgerInterceptor()).getIndex();
+    }
+
+    public static CompletableFuture<Long> getOffsetOfPosition(ManagedLedgerImpl managedLedger, PositionImpl position) {
+        final CompletableFuture<Long> future = new CompletableFuture<>();
+        managedLedger.asyncReadEntry(position, new AsyncCallbacks.ReadEntryCallback() {
+            @Override
+            public void readEntryFailed(ManagedLedgerException exception, Object ctx) {
+                future.completeExceptionally(exception);
+            }
+
+            @Override
+            public void readEntryComplete(Entry entry, Object ctx) {
+                try {
+                    PulsarApi.BrokerEntryMetadata entryMetadata =
+                            Commands.parseBrokerEntryMetadataIfExist(entry.getDataBuffer());
+                   future.complete(entryMetadata.getIndex());
+                } catch (Exception exception) {
+                    future.completeExceptionally(exception);
+                } finally {
+                    if (entry != null) {
+                        entry.release();
+                    }
+                }
+            }
+        }, null);
+        return future;
+    }
+
+    public static PositionImpl getPositionForOffset(ManagedLedger managedLedger, Long offset) {
+        try {
+            return (PositionImpl) managedLedger.asyncFindPosition(new OffsetSearchPredicate(offset)).get();
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("[{}] Failed to find position for offset {}", managedLedger.getName(), offset);
+            throw new RuntimeException(managedLedger.getName() + " failed to find position for offset " + offset);
+        }
+    }
+
+    public static PulsarApi.BrokerEntryMetadata peekBrokerEntryMetadata(ByteBuf byteBuf) {
+        final int readerIndex = byteBuf.readerIndex();
+        PulsarApi.BrokerEntryMetadata entryMetadata =
+                Commands.parseBrokerEntryMetadataIfExist(byteBuf);
+        byteBuf.readerIndex(readerIndex);
+        return entryMetadata;
+    }
+
+    public static long peekOffsetFromEntry(Entry entry) {
+        return peekBrokerEntryMetadata(entry.getDataBuffer()).getIndex();
+    }
+
+    public static long peekBaseOffsetFromEntry(Entry entry) {
+
+        return peekOffsetFromEntry(entry)
+                - Commands.peekMessageMetadata(entry.getDataBuffer(), null, 0)
+                    .getNumMessagesInBatch() + 1;
+    }
+
+    public static long getMockOffset(long ledgerId, long entryId) {
+        return ledgerId + entryId;
+    }
```

Comment:
Could you explain what's the meaning of `ledgerId + entryId`? I found `MessagePublishContext#complete` use it as the offset which would be sent back to the producer.

Reply:
This a mock method to provide a simulate offset after append one entry successfully. It's invoked from two places :
1. from `MessagePublishContext` when append entry to bookie 
2. from `GroupMetadataManager` when  commited offset write to bookie
as far as I can see, the only requirement for this method is this method should generate a monotone increasing result.

Reply:
You're probably right. I also think a monotonically increasing offset is enough. If existed tests could pass, I think the precise offset could be implemented later.